### PR TITLE
Added type check for function calls in templates

### PR DESCRIPTION
**Pull request by ansharlubis**
Clean PR for #1055.

### CODE REVIEW

File: tests/errors/template_error.f90 [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+end module
+
+program template_travel
+use template_travel_m
+implicit none
+
+call test_template()
+
+end program template_travel
```
[the remark this answers is not in the document]
Comment (czgdp1807) [reply tests/errors/template_error.f90:109]: We will merge both of your LPython (the one you will create after #1166 is merged) and LFortran PR (this one) together. Until then marking this as draft.

File: tests/reference/asr-template_error-8ad11a2.stderr [hunk 2]
```diff
@@ -0,0 +1,12 @@
+semantic error: Type mismatch in function call, the types must be compatible
```
Comment (certik) [inline tests/reference/asr-template_error-8ad11a2.stderr:1]:
```suggestion
semantic error: Type mismatch in function call, the function expects t, but s was provided
```

File: src/lfortran/semantics/ast_common_visitor.h [hunk 3]
```diff
@@ -3392,9 +3392,32 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
...
+                        ASR::ttype_t *source_type = ASRUtils::expr_type(source);
+
+                        if (!ASRUtils::check_equal_type(dest_type, source_type)) {
+                            std::string dtype = ASRUtils::type_to_str(dest_type);
+                            std::string stype = ASRUtils::type_to_str(source_type);
+                            diag.add(Diagnostic(
+                                "Type mismatch in function call, the types must be compatible",
+                                Level::Error, Stage::Semantic, {
+                                    Label("type mismatch (" + dtype + " and " + stype + ")",
+                                            {dest->base.loc, source->base.loc})
```
[the remark this answers is not in the document]
Comment (certik) [reply src/lfortran/semantics/ast_common_visitor.h:3413]: Underneath I would copy:
```
                                    Label("function definition has type `" + dtype + "`",
                                            {dest->base.loc})
```
The `dest` should point to the argument only, but for some reason it seems to point to the whole function, so the location information might be wrong, but that's a separate issue.

**Review by certik — APPROVED** [review]
I think this PR works, so we can merge it as is (squash).

Then you can submit a new PR to improve the error message, based on my comments above.

**Review by czgdp1807 — APPROVED** [review]

**Pull request merged**